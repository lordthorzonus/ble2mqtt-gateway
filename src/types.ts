import { EnhancedRuuviTagSensorData } from "./gateways/ruuvitag/ruuvitag-sensor-data-decorator";
import { DateTime } from "luxon";
import { MiFloraSensorData } from "./gateways/miflora/miflora-measurement-transformer";

export enum DeviceType {
    Ruuvitag = "ruuvitag",
    MiFlora = "miflora",
}

export enum MessageType {
    Availability = "availability",
    SensorData = "sensor-data",
    Analytics = "analytics",
}

export interface Device {
    id: string;
    friendlyName: string;
    type: DeviceType;
}

export interface DeviceSensorMessage {
    device: {
        macAddress: string;
        id: string;
        rssi: number | null;
        friendlyName: string;
        type: DeviceType;
        timeout: number;
    };
    id: string;
    time: DateTime;
    type: MessageType.SensorData;
    payload: Record<string, unknown>;
}

export interface DeviceAvailabilityMessage {
    device: {
        macAddress: string;
        id: string;
        rssi: number | null;
        friendlyName: string;
        type: DeviceType;
        timeout: number;
    };
    id: string;
    time: DateTime;
    type: MessageType.Availability;
    payload: {
        state: "online" | "offline";
    };
}

export type DeviceMessage = DeviceAvailabilityMessage | DeviceSensorMessage;
export type BleGatewayMessage = DeviceMessage;

export interface MqttMessage {
    topic: string;
    payload: string;
    retain: boolean;
}

export enum HomeAssistantDeviceClass {
    Temperature = "temperature",
    Humidity = "humidity",
    Pressure = "pressure",
    Battery = "battery",
    None = "none",
    Illuminance = "illuminance",
}

export enum HomeAssistantMQTTComponent {
    Sensor = "sensor",
    BinarySensor = "binary_sensor",
}

export interface HomeAssistantSensorConfiguration {
    name: string;
    deviceClass: HomeAssistantDeviceClass;
    unitOfMeasurement: string;
    valueTemplate?: string;
    uniqueId: string;
    icon?: string;
    stateClass?: string;
    suggestedDecimalPrecision?: number | null;
    device: {
        manufacturer: string;
        model: string;
    };
}

export type HomeAssistantSensorConfigurationForDevice<T> =
    | {
          [K in keyof T]: HomeAssistantSensorConfiguration;
      }
    | Record<string, HomeAssistantSensorConfiguration & { valueTemplate: string }>;

export type HomeAssistantSensorConfigurationForDeviceType<T> = T extends DeviceType.Ruuvitag
    ? HomeAssistantSensorConfigurationForDevice<EnhancedRuuviTagSensorData>
    : T extends DeviceType.MiFlora
    ? HomeAssistantSensorConfigurationForDevice<MiFloraSensorData>
    : never;
