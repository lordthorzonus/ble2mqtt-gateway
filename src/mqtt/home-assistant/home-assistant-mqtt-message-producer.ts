import {
    DeviceAvailabilityMessage,
    DeviceMessage,
    DeviceType,
    HomeAssistantDeviceClass,
    HomeAssistantMQTTComponent,
    HomeAssistantSensorConfiguration,
    HomeAssistantSensorConfigurationForDeviceType,
    MessageType,
    MqttMessage,
} from "../../types";
import { ruuviTagSensorConfiguration } from "./device-discoverability-configurations/ruuvitag";
import { getDeviceAvailabilityTopic, getDeviceStateTopic } from "../mqtt-topic-factory";
import { Observable } from "rxjs";
import { getConfiguration } from "../../config";
import { miFloraSensorConfiguration } from "./device-discoverability-configurations/miflora";
import { snakeCase } from "lodash";

const config = getConfiguration();
const homeAssistantTopicBase = config.homeassistant.discovery_topic;

const bleDeviceHAConfigurationMap = new Map<DeviceType, HomeAssistantSensorConfigurationForDeviceType<DeviceType>>([
    [DeviceType.Ruuvitag, ruuviTagSensorConfiguration],
    [DeviceType.MiFlora, miFloraSensorConfiguration],
]);

const bleDeviceHAComponentMap = new Map<DeviceType, HomeAssistantMQTTComponent>([
    [DeviceType.Ruuvitag, HomeAssistantMQTTComponent.Sensor],
    [DeviceType.MiFlora, HomeAssistantMQTTComponent.Sensor],
]);

const getHASensorConfiguration = <T extends DeviceType>(type: T): HomeAssistantSensorConfigurationForDeviceType<T> => {
    const configuration = bleDeviceHAConfigurationMap.get(type);

    if (!configuration) {
        throw new Error("No configuration for given type");
    }

    return configuration as HomeAssistantSensorConfigurationForDeviceType<T>;
};

const getHAComponent = (type: DeviceType): HomeAssistantMQTTComponent => {
    const component = bleDeviceHAComponentMap.get(type);

    if (!component) {
        throw new Error("No mqtt component configured for given type");
    }

    return component;
};

const getObjectID = (deviceMessage: DeviceMessage, configEntry: HomeAssistantSensorConfiguration) =>
    `${snakeCase(deviceMessage.device.friendlyName)}_${configEntry.uniqueId}`;

const getDeviceName = (deviceMessage: DeviceMessage) =>
    `${deviceMessage.device.type} ${deviceMessage.device.friendlyName}`;

const getEntityName = (deviceMessage: DeviceMessage, configEntry: HomeAssistantSensorConfiguration) =>
    `${deviceMessage.device.friendlyName} ${configEntry.name}`;

interface HADiscoveryPayload {
    name: string;
    device_class?: string;
    expire_after: number;
    unit_of_measurement: string;
    object_id: string;
    unique_id: string;
    availability_topic: string;
    availability_template: string;
    value_template: string;
    state_topic: string;
    icon?: string;
    device: {
        name: string;
        manufacturer: string;
        model: string;
        connections: [string, string][];
    };
}

const getHaDiscoveryPayload = (
    propertyName: string,
    configEntry: HomeAssistantSensorConfiguration,
    deviceMessage: DeviceAvailabilityMessage
): HADiscoveryPayload | null =>
    deviceMessage.payload.state === "online"
        ? {
              name: getEntityName(deviceMessage, configEntry),
              device_class:
                  configEntry.deviceClass === HomeAssistantDeviceClass.None ? undefined : configEntry.deviceClass,
              unit_of_measurement: configEntry.unitOfMeasurement,
              object_id: getObjectID(deviceMessage, configEntry),
              unique_id: getObjectID(deviceMessage, configEntry),
              availability_topic: getDeviceAvailabilityTopic(deviceMessage.device),
              availability_template: "{{ value_json.state }}",
              expire_after: deviceMessage.device.timeout / 1000,
              icon: configEntry.icon,
              device: {
                  ...configEntry.device,
                  name: getDeviceName(deviceMessage),
                  connections: [["mac", deviceMessage.device.macAddress]],
              },
              value_template: configEntry.valueTemplate
                  ? configEntry.valueTemplate
                  : `{{ value_json.${propertyName} }}`,
              state_topic: getDeviceStateTopic(deviceMessage.device),
          }
        : null;

function* haDiscoverAdapter(deviceMessage: DeviceAvailabilityMessage): Generator<MqttMessage> {
    const configuration = getHASensorConfiguration(deviceMessage.device.type);
    const component = getHAComponent(deviceMessage.device.type);

    for (const [propertyName, configEntry] of Object.entries(configuration)) {
        const haDiscoveryPayload = getHaDiscoveryPayload(propertyName, configEntry, deviceMessage);

        const message: MqttMessage = {
            topic: `${homeAssistantTopicBase}/${component}/${getObjectID(deviceMessage, configEntry)}/config`,
            retain: true,
            payload: haDiscoveryPayload !== null ? JSON.stringify(haDiscoveryPayload) : "",
        };

        yield message;
    }
}

const generateAvailabilityMessage = (deviceMessage: DeviceMessage): MqttMessage => ({
    retain: true,
    topic: getDeviceAvailabilityTopic(deviceMessage.device),
    payload: JSON.stringify(deviceMessage.payload),
});

const isAvailabilityMessage = (deviceMessage: DeviceMessage): deviceMessage is DeviceAvailabilityMessage =>
    deviceMessage.type === MessageType.Availability;

const generateStateMessage = (deviceMessage: DeviceMessage): MqttMessage => ({
    retain: false,
    topic: getDeviceStateTopic(deviceMessage.device),
    payload: JSON.stringify({ ...deviceMessage.payload, time: deviceMessage.time, id: deviceMessage.id }),
});

export function homeAssistantMqttMessageProducer(deviceMessage: DeviceMessage): Observable<MqttMessage> {
    return new Observable((subscriber) => {
        if (isAvailabilityMessage(deviceMessage)) {
            for (const message of haDiscoverAdapter(deviceMessage)) {
                subscriber.next(message);
            }

            subscriber.next(generateAvailabilityMessage(deviceMessage));
            subscriber.complete();
        }

        subscriber.next(generateStateMessage(deviceMessage));
        subscriber.complete();
    });
}
