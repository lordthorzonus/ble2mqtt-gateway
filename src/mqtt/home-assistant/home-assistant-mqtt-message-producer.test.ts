import { toArray } from "rxjs";

jest.mock("../../config", () => ({
    __esModule: true,
    getConfiguration: jest.fn().mockReturnValue({
        decimal_precision: 2,
        gateways: {
            base_topic: "a_base_topic",
        },
        homeassistant: {
            discovery_topic: "homeassistant",
        },
    }),
}));

import { DeviceAvailabilityMessage, DeviceSensorMessage, MessageType, DeviceType } from "../../types";
import { DateTime } from "luxon";
import { homeAssistantMqttMessageProducer } from "./home-assistant-mqtt-message-producer";

describe("HomeAssistant MQTT Message producer", () => {
    it("should generate device state mqtt message if device sensor message is given", (done) => {
        const deviceMessage: DeviceSensorMessage = {
            id: "a",
            payload: {
                sensor1: "value",
            },
            type: MessageType.SensorData,
            time: DateTime.now(),
            device: {
                macAddress: "aa:bb",
                id: "aa:bb",
                friendlyName: "Fridge friend",
                rssi: 23,
                type: DeviceType.Ruuvitag,
                timeout: 10000,
            },
        };

        homeAssistantMqttMessageProducer(deviceMessage).subscribe((message) => {
            expect(message).toEqual({
                payload: JSON.stringify({
                    sensor1: "value",
                    time: deviceMessage.time,
                    id: deviceMessage.id,
                }),
                retain: false,
                topic: "a_base_topic/ruuvitag/aa:bb/state",
            });
            done();
        });
    });

    it("should vary the state topic based on the device type", (done) => {
        const deviceMessage: DeviceSensorMessage = {
            id: "a",
            payload: {
                sensor1: "value",
            },
            type: MessageType.SensorData,
            time: DateTime.now(),
            device: {
                macAddress: "aa:bb",
                id: "aa:bb",
                friendlyName: "My Plant",
                rssi: 44,
                type: DeviceType.MiFlora,
                timeout: 10000,
            },
        };

        homeAssistantMqttMessageProducer(deviceMessage).subscribe((message) => {
            expect(message).toEqual({
                payload: JSON.stringify({
                    sensor1: "value",
                    time: deviceMessage.time,
                    id: deviceMessage.id,
                }),
                retain: false,
                topic: "a_base_topic/miflora/aa:bb/state",
            });
            done();
        });
    });

    const availabilityTestCases: [string, DeviceAvailabilityMessage][] = [
        [
            "RuuviTag Discovered",
            {
                id: "a",
                time: DateTime.now(),
                payload: { state: "online" },
                type: MessageType.Availability,
                device: {
                    macAddress: "aa:bb",
                    id: "aa:bb",
                    friendlyName: "My Fridge",
                    rssi: 23,
                    type: DeviceType.Ruuvitag,
                    timeout: 10000,
                },
            },
        ],
        [
            "RuuviTag Offline",
            {
                id: "a",
                time: DateTime.now(),
                payload: { state: "offline" },
                type: MessageType.Availability,
                device: {
                    macAddress: "aa:bb",
                    id: "aa:bb",
                    friendlyName: "fridge",
                    rssi: 23,
                    type: DeviceType.Ruuvitag,
                    timeout: 30000,
                },
            },
        ],
        [
            "MiFlora Discovered",
            {
                id: "abba",
                time: DateTime.now(),
                payload: { state: "online" },
                type: MessageType.Availability,
                device: {
                    macAddress: "aa:bb",
                    id: "aa:bb",
                    friendlyName: "Super Plant",
                    rssi: 23,
                    type: DeviceType.MiFlora,
                    timeout: 30000,
                },
            },
        ],
        [
            "MiFlora Offline",
            {
                id: "abba",
                time: DateTime.now(),
                payload: { state: "offline" },
                type: MessageType.Availability,
                device: {
                    macAddress: "aa:bb",
                    id: "aa:bb",
                    friendlyName: "plant",
                    rssi: 23,
                    type: DeviceType.MiFlora,
                    timeout: 30000,
                },
            },
        ],
    ];
    it.each(availabilityTestCases)(
        "should generate corresponding discovery messages for given device in case of %s availability message is given",
        (_, availabilityMessage) => {
            homeAssistantMqttMessageProducer(availabilityMessage)
                .pipe(toArray())
                .subscribe((messages) => {
                    expect(messages).toMatchSnapshot();
                });
        }
    );
});
