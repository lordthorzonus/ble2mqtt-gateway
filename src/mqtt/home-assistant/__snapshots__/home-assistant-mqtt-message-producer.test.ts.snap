// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HomeAssistant MQTT Message producer should generate corresponding discovery messages for given device in case of RuuviTag Discovered availability message is given 1`] = `
Array [
  Object {
    "payload": "{\\"name\\":\\"fridge Absolute Humidity\\",\\"unit_of_measurement\\":\\"g/m^3\\",\\"object_id\\":\\"fridge_absolute_humidity\\",\\"unique_id\\":\\"fridge_absolute_humidity\\",\\"availability_topic\\":\\"a_base_topic/ruuvitag/aa:bb/availability\\",\\"availability_template\\":\\"{{ value_json.state }}\\",\\"expire_after\\":10,\\"device\\":{\\"manufacturer\\":\\"Ruuvi Innovations Oy\\",\\"model\\":\\"RuuviTag\\",\\"name\\":\\"ruuvitag fridge\\",\\"connections\\":[[\\"mac\\",\\"aa:bb\\"]]},\\"value_template\\":\\"{{ value_json.absoluteHumidity }}\\",\\"state_topic\\":\\"a_base_topic/ruuvitag/aa:bb/state\\"}",
    "retain": true,
    "topic": "homeassistant/sensor/fridge_absolute_humidity/config",
  },
  Object {
    "payload": "{\\"name\\":\\"fridge Temperature\\",\\"device_class\\":\\"temperature\\",\\"unit_of_measurement\\":\\"°C\\",\\"object_id\\":\\"fridge_temperature\\",\\"unique_id\\":\\"fridge_temperature\\",\\"availability_topic\\":\\"a_base_topic/ruuvitag/aa:bb/availability\\",\\"availability_template\\":\\"{{ value_json.state }}\\",\\"expire_after\\":10,\\"device\\":{\\"manufacturer\\":\\"Ruuvi Innovations Oy\\",\\"model\\":\\"RuuviTag\\",\\"name\\":\\"ruuvitag fridge\\",\\"connections\\":[[\\"mac\\",\\"aa:bb\\"]]},\\"value_template\\":\\"{{ value_json.temperature }}\\",\\"state_topic\\":\\"a_base_topic/ruuvitag/aa:bb/state\\"}",
    "retain": true,
    "topic": "homeassistant/sensor/fridge_temperature/config",
  },
  Object {
    "payload": "{\\"name\\":\\"fridge Battery Voltage\\",\\"unit_of_measurement\\":\\"mV\\",\\"object_id\\":\\"fridge_battery_voltage\\",\\"unique_id\\":\\"fridge_battery_voltage\\",\\"availability_topic\\":\\"a_base_topic/ruuvitag/aa:bb/availability\\",\\"availability_template\\":\\"{{ value_json.state }}\\",\\"expire_after\\":10,\\"icon\\":\\"mdi:battery\\",\\"device\\":{\\"manufacturer\\":\\"Ruuvi Innovations Oy\\",\\"model\\":\\"RuuviTag\\",\\"name\\":\\"ruuvitag fridge\\",\\"connections\\":[[\\"mac\\",\\"aa:bb\\"]]},\\"value_template\\":\\"{{ value_json.batteryVoltage }}\\",\\"state_topic\\":\\"a_base_topic/ruuvitag/aa:bb/state\\"}",
    "retain": true,
    "topic": "homeassistant/sensor/fridge_battery_voltage/config",
  },
  Object {
    "payload": "{\\"name\\":\\"fridge Humidity\\",\\"device_class\\":\\"humidity\\",\\"unit_of_measurement\\":\\"%\\",\\"object_id\\":\\"fridge_humidity\\",\\"unique_id\\":\\"fridge_humidity\\",\\"availability_topic\\":\\"a_base_topic/ruuvitag/aa:bb/availability\\",\\"availability_template\\":\\"{{ value_json.state }}\\",\\"expire_after\\":10,\\"icon\\":\\"mdi:gauge\\",\\"device\\":{\\"manufacturer\\":\\"Ruuvi Innovations Oy\\",\\"model\\":\\"RuuviTag\\",\\"name\\":\\"ruuvitag fridge\\",\\"connections\\":[[\\"mac\\",\\"aa:bb\\"]]},\\"value_template\\":\\"{{ value_json.relativeHumidityPercentage | float(2) }}\\",\\"state_topic\\":\\"a_base_topic/ruuvitag/aa:bb/state\\"}",
    "retain": true,
    "topic": "homeassistant/sensor/fridge_humidity/config",
  },
  Object {
    "payload": "{\\"name\\":\\"fridge Dew Point\\",\\"unit_of_measurement\\":\\"°C\\",\\"object_id\\":\\"fridge_dew_point\\",\\"unique_id\\":\\"fridge_dew_point\\",\\"availability_topic\\":\\"a_base_topic/ruuvitag/aa:bb/availability\\",\\"availability_template\\":\\"{{ value_json.state }}\\",\\"expire_after\\":10,\\"device\\":{\\"manufacturer\\":\\"Ruuvi Innovations Oy\\",\\"model\\":\\"RuuviTag\\",\\"name\\":\\"ruuvitag fridge\\",\\"connections\\":[[\\"mac\\",\\"aa:bb\\"]]},\\"value_template\\":\\"{{ value_json.dewPoint }}\\",\\"state_topic\\":\\"a_base_topic/ruuvitag/aa:bb/state\\"}",
    "retain": true,
    "topic": "homeassistant/sensor/fridge_dew_point/config",
  },
  Object {
    "payload": "{\\"name\\":\\"fridge Pressure\\",\\"device_class\\":\\"pressure\\",\\"unit_of_measurement\\":\\"hPa\\",\\"object_id\\":\\"fridge_pressure\\",\\"unique_id\\":\\"fridge_pressure\\",\\"availability_topic\\":\\"a_base_topic/ruuvitag/aa:bb/availability\\",\\"availability_template\\":\\"{{ value_json.state }}\\",\\"expire_after\\":10,\\"device\\":{\\"manufacturer\\":\\"Ruuvi Innovations Oy\\",\\"model\\":\\"RuuviTag\\",\\"name\\":\\"ruuvitag fridge\\",\\"connections\\":[[\\"mac\\",\\"aa:bb\\"]]},\\"value_template\\":\\"{{ value_json.pressure }}\\",\\"state_topic\\":\\"a_base_topic/ruuvitag/aa:bb/state\\"}",
    "retain": true,
    "topic": "homeassistant/sensor/fridge_pressure/config",
  },
  Object {
    "payload": "{\\"name\\":\\"fridge TX Power\\",\\"unit_of_measurement\\":\\"dBm\\",\\"object_id\\":\\"fridge_txPower\\",\\"unique_id\\":\\"fridge_txPower\\",\\"availability_topic\\":\\"a_base_topic/ruuvitag/aa:bb/availability\\",\\"availability_template\\":\\"{{ value_json.state }}\\",\\"expire_after\\":10,\\"icon\\":\\"mdi:signal\\",\\"device\\":{\\"manufacturer\\":\\"Ruuvi Innovations Oy\\",\\"model\\":\\"RuuviTag\\",\\"name\\":\\"ruuvitag fridge\\",\\"connections\\":[[\\"mac\\",\\"aa:bb\\"]]},\\"value_template\\":\\"{{ value_json.txPower }}\\",\\"state_topic\\":\\"a_base_topic/ruuvitag/aa:bb/state\\"}",
    "retain": true,
    "topic": "homeassistant/sensor/fridge_txPower/config",
  },
  Object {
    "payload": "{\\"name\\":\\"fridge Acceleration X\\",\\"unit_of_measurement\\":\\"mG\\",\\"object_id\\":\\"fridge_acceleration_x\\",\\"unique_id\\":\\"fridge_acceleration_x\\",\\"availability_topic\\":\\"a_base_topic/ruuvitag/aa:bb/availability\\",\\"availability_template\\":\\"{{ value_json.state }}\\",\\"expire_after\\":10,\\"device\\":{\\"manufacturer\\":\\"Ruuvi Innovations Oy\\",\\"model\\":\\"RuuviTag\\",\\"name\\":\\"ruuvitag fridge\\",\\"connections\\":[[\\"mac\\",\\"aa:bb\\"]]},\\"value_template\\":\\"{{ value_json.accelerationX }}\\",\\"state_topic\\":\\"a_base_topic/ruuvitag/aa:bb/state\\"}",
    "retain": true,
    "topic": "homeassistant/sensor/fridge_acceleration_x/config",
  },
  Object {
    "payload": "{\\"name\\":\\"fridge Acceleration Y\\",\\"unit_of_measurement\\":\\"mG\\",\\"object_id\\":\\"fridge_acceleration_y\\",\\"unique_id\\":\\"fridge_acceleration_y\\",\\"availability_topic\\":\\"a_base_topic/ruuvitag/aa:bb/availability\\",\\"availability_template\\":\\"{{ value_json.state }}\\",\\"expire_after\\":10,\\"device\\":{\\"manufacturer\\":\\"Ruuvi Innovations Oy\\",\\"model\\":\\"RuuviTag\\",\\"name\\":\\"ruuvitag fridge\\",\\"connections\\":[[\\"mac\\",\\"aa:bb\\"]]},\\"value_template\\":\\"{{ value_json.accelerationY }}\\",\\"state_topic\\":\\"a_base_topic/ruuvitag/aa:bb/state\\"}",
    "retain": true,
    "topic": "homeassistant/sensor/fridge_acceleration_y/config",
  },
  Object {
    "payload": "{\\"name\\":\\"fridge Acceleration Z\\",\\"unit_of_measurement\\":\\"mG\\",\\"object_id\\":\\"fridge_acceleration_z\\",\\"unique_id\\":\\"fridge_acceleration_z\\",\\"availability_topic\\":\\"a_base_topic/ruuvitag/aa:bb/availability\\",\\"availability_template\\":\\"{{ value_json.state }}\\",\\"expire_after\\":10,\\"device\\":{\\"manufacturer\\":\\"Ruuvi Innovations Oy\\",\\"model\\":\\"RuuviTag\\",\\"name\\":\\"ruuvitag fridge\\",\\"connections\\":[[\\"mac\\",\\"aa:bb\\"]]},\\"value_template\\":\\"{{ value_json.accelerationZ }}\\",\\"state_topic\\":\\"a_base_topic/ruuvitag/aa:bb/state\\"}",
    "retain": true,
    "topic": "homeassistant/sensor/fridge_acceleration_z/config",
  },
  Object {
    "payload": "{\\"name\\":\\"fridge Humidex\\",\\"unit_of_measurement\\":\\"\\",\\"object_id\\":\\"fridge_humidex\\",\\"unique_id\\":\\"fridge_humidex\\",\\"availability_topic\\":\\"a_base_topic/ruuvitag/aa:bb/availability\\",\\"availability_template\\":\\"{{ value_json.state }}\\",\\"expire_after\\":10,\\"device\\":{\\"manufacturer\\":\\"Ruuvi Innovations Oy\\",\\"model\\":\\"RuuviTag\\",\\"name\\":\\"ruuvitag fridge\\",\\"connections\\":[[\\"mac\\",\\"aa:bb\\"]]},\\"value_template\\":\\"{{ value_json.humidex }}\\",\\"state_topic\\":\\"a_base_topic/ruuvitag/aa:bb/state\\"}",
    "retain": true,
    "topic": "homeassistant/sensor/fridge_humidex/config",
  },
  Object {
    "payload": "{\\"name\\":\\"fridge Heat Index\\",\\"unit_of_measurement\\":\\"°C\\",\\"object_id\\":\\"fridge_heat_index\\",\\"unique_id\\":\\"fridge_heat_index\\",\\"availability_topic\\":\\"a_base_topic/ruuvitag/aa:bb/availability\\",\\"availability_template\\":\\"{{ value_json.state }}\\",\\"expire_after\\":10,\\"device\\":{\\"manufacturer\\":\\"Ruuvi Innovations Oy\\",\\"model\\":\\"RuuviTag\\",\\"name\\":\\"ruuvitag fridge\\",\\"connections\\":[[\\"mac\\",\\"aa:bb\\"]]},\\"value_template\\":\\"{{ value_json.heatIndex }}\\",\\"state_topic\\":\\"a_base_topic/ruuvitag/aa:bb/state\\"}",
    "retain": true,
    "topic": "homeassistant/sensor/fridge_heat_index/config",
  },
  Object {
    "payload": "{\\"name\\":\\"fridge Mac Address\\",\\"unit_of_measurement\\":\\"\\",\\"object_id\\":\\"fridge_mac_address\\",\\"unique_id\\":\\"fridge_mac_address\\",\\"availability_topic\\":\\"a_base_topic/ruuvitag/aa:bb/availability\\",\\"availability_template\\":\\"{{ value_json.state }}\\",\\"expire_after\\":10,\\"device\\":{\\"manufacturer\\":\\"Ruuvi Innovations Oy\\",\\"model\\":\\"RuuviTag\\",\\"name\\":\\"ruuvitag fridge\\",\\"connections\\":[[\\"mac\\",\\"aa:bb\\"]]},\\"value_template\\":\\"{{ value_json.macAddress }}\\",\\"state_topic\\":\\"a_base_topic/ruuvitag/aa:bb/state\\"}",
    "retain": true,
    "topic": "homeassistant/sensor/fridge_mac_address/config",
  },
  Object {
    "payload": "{\\"name\\":\\"fridge Measurement Sequence\\",\\"unit_of_measurement\\":\\"\\",\\"object_id\\":\\"fridge_measurement_sequence\\",\\"unique_id\\":\\"fridge_measurement_sequence\\",\\"availability_topic\\":\\"a_base_topic/ruuvitag/aa:bb/availability\\",\\"availability_template\\":\\"{{ value_json.state }}\\",\\"expire_after\\":10,\\"device\\":{\\"manufacturer\\":\\"Ruuvi Innovations Oy\\",\\"model\\":\\"RuuviTag\\",\\"name\\":\\"ruuvitag fridge\\",\\"connections\\":[[\\"mac\\",\\"aa:bb\\"]]},\\"value_template\\":\\"{{ value_json.measurementSequence }}\\",\\"state_topic\\":\\"a_base_topic/ruuvitag/aa:bb/state\\"}",
    "retain": true,
    "topic": "homeassistant/sensor/fridge_measurement_sequence/config",
  },
  Object {
    "payload": "{\\"name\\":\\"fridge Movement Counter\\",\\"unit_of_measurement\\":\\"\\",\\"object_id\\":\\"fridge_movement_counter\\",\\"unique_id\\":\\"fridge_movement_counter\\",\\"availability_topic\\":\\"a_base_topic/ruuvitag/aa:bb/availability\\",\\"availability_template\\":\\"{{ value_json.state }}\\",\\"expire_after\\":10,\\"device\\":{\\"manufacturer\\":\\"Ruuvi Innovations Oy\\",\\"model\\":\\"RuuviTag\\",\\"name\\":\\"ruuvitag fridge\\",\\"connections\\":[[\\"mac\\",\\"aa:bb\\"]]},\\"value_template\\":\\"{{ value_json.movementCounter }}\\",\\"state_topic\\":\\"a_base_topic/ruuvitag/aa:bb/state\\"}",
    "retain": true,
    "topic": "homeassistant/sensor/fridge_movement_counter/config",
  },
  Object {
    "payload": "{\\"state\\":\\"online\\"}",
    "retain": true,
    "topic": "a_base_topic/ruuvitag/aa:bb/availability",
  },
]
`;

exports[`HomeAssistant MQTT Message producer should generate corresponding discovery messages for given device in case of RuuviTag Offline availability message is given 1`] = `
Array [
  Object {
    "payload": "null",
    "retain": true,
    "topic": "homeassistant/sensor/fridge_absolute_humidity/config",
  },
  Object {
    "payload": "null",
    "retain": true,
    "topic": "homeassistant/sensor/fridge_temperature/config",
  },
  Object {
    "payload": "null",
    "retain": true,
    "topic": "homeassistant/sensor/fridge_battery_voltage/config",
  },
  Object {
    "payload": "null",
    "retain": true,
    "topic": "homeassistant/sensor/fridge_humidity/config",
  },
  Object {
    "payload": "null",
    "retain": true,
    "topic": "homeassistant/sensor/fridge_dew_point/config",
  },
  Object {
    "payload": "null",
    "retain": true,
    "topic": "homeassistant/sensor/fridge_pressure/config",
  },
  Object {
    "payload": "null",
    "retain": true,
    "topic": "homeassistant/sensor/fridge_txPower/config",
  },
  Object {
    "payload": "null",
    "retain": true,
    "topic": "homeassistant/sensor/fridge_acceleration_x/config",
  },
  Object {
    "payload": "null",
    "retain": true,
    "topic": "homeassistant/sensor/fridge_acceleration_y/config",
  },
  Object {
    "payload": "null",
    "retain": true,
    "topic": "homeassistant/sensor/fridge_acceleration_z/config",
  },
  Object {
    "payload": "null",
    "retain": true,
    "topic": "homeassistant/sensor/fridge_humidex/config",
  },
  Object {
    "payload": "null",
    "retain": true,
    "topic": "homeassistant/sensor/fridge_heat_index/config",
  },
  Object {
    "payload": "null",
    "retain": true,
    "topic": "homeassistant/sensor/fridge_mac_address/config",
  },
  Object {
    "payload": "null",
    "retain": true,
    "topic": "homeassistant/sensor/fridge_measurement_sequence/config",
  },
  Object {
    "payload": "null",
    "retain": true,
    "topic": "homeassistant/sensor/fridge_movement_counter/config",
  },
  Object {
    "payload": "{\\"state\\":\\"offline\\"}",
    "retain": true,
    "topic": "a_base_topic/ruuvitag/aa:bb/availability",
  },
]
`;
